First draft of Auth0 + React (#266)
MechanicalRock/mechanicalrock.github.io: _posts/2022-12-06-integrating-auth0-with-react.md

---
layout: post
font: serif
title: Integrating Auth0 Authentication into React

> date: 2022-12-06

Reviewer: Note, you may want to update this prior to publish.  It must be a data in the past at the point of merge though

highlight: monokai
author: Chris Howard
image: img/blog/integrating-auth0-with-react/banner.jpg
tags: Auth0 react authentication typescript
description: 
---

> Authentication and Authorization is objectively the most important part of any application, So why is it so often left to the end of development cycles or implemented in a slap dash manner?

Reviewer: > Authentication and Authorization

You only cover Authentication in this post.  

Do you plan to do a follow up for authorization (which would be great)?

They are long enough to split into separate posts, but you should call out that you are only covering one half somewhere.  Either in the introduction or outro.

> It may be due to the difficulty in creating authentication systems and implementing them into our applications, but with the tools and services available to developers it has become easier and easier to create a robust, secure and easy to use system. So let's not wait and get started on integrating [Auth0](https://auth0.com/) into a new React application!

Reviewer: Nice hook! :+1:

<br />

## Contents

- [Why Auth0?](#why-auth0)
- [Create Auth0 Application](#create-auth0-application)
- [Integration into React](#integrate-with-react)
    * [Login](#but-wait-how-do-users-login)
    * [Logout](#great-now-how-about-logout)
- [Conclusion](#conclusion)

> ## Why Auth0?

Reviewer: You don't introduce OAuth.

Do you expect the reader to understand about OAuth?  Is it pre-requisite knowleddge to this post?

You probably need a short sentence/introduction at least, with a link to other reference documentation for pople to learn more.

Author: Generally not a huge amount of knowledge regarding OAuth is strictly required for just the basic implementation, Was worried it would get too long

Will add a brief touch on it and a link

> Auth0 is one of the most highly reputed company working in this space. Why build your own service when you can get the years of experience and knowledge dedicated to making sure your users login data is safe?

Reviewer: ```suggestion
Auth0 is one of the most highly reputed companies working in this space. Why build your own service when you can get the years of experience and knowledge dedicated to making sure your users login data is safe?
```

tpoy

Auth0 provides Authentication, Authorization, Social Logins, MFA, Analytics and lots more, And to make it all the sweeter, has amazing developer documentation. All of this completely free if you have under 7000 users!

## Create Auth0 Application

> Let's start with create a new Auth0 Application

Reviewer: > Let's start with create a new Auth0 Application

grammar.  Not sure what you're trying to say.

Author: Fixed spelling

> Let's start with creating a new Auth0 Application

Reviewer: ```suggestion
Let's start creating a new Auth0 Application
```

Or

```suggestion
Let's start by creating a new Auth0 Application
```

Not sure how well GitHub will deal with alternatives :)

> Select Applications > Applications from the side menu (or alternatively the Create Application on the Getting Started page)

Reviewer: You're missing context here = where should I be?

In an application?  
On a web page?  Which one?

Author: Added some additional context

> Login to [Auth0](https://auth0.com/) (or create a new account if you haven't already) and navigate to your [Dashboard](https://manage.auth0.com/dashboard)

> ![Create Application](/img/blog/integrating-auth0-with-react/create_application.png){:width="962px" height="173px" loading="lazy"}

Reviewer 2: Good job on using the Lazy attribute ❤️

![Create Application Getting Started](/img/blog/integrating-auth0-with-react/create_application_alt.png){:width="797px" height="277px" loading="lazy"}

Here we will be able to give our Auth0 Application a name, and a type, In our example, we will select **Single Page Web Applications**

![Create Application Popup](/img/blog/integrating-auth0-with-react/create_application_popup.png){:width="600px" height="543px" loading="lazy"}

Once the application has been created, Navigate to the Settings and these are the key credentials we will need to integrate Auth0 into our application

![Application Settings](/img/blog/integrating-auth0-with-react/client_credentials.png){:width="760px" height="330px" loading="lazy"}

> One last step before we can begin integrating! We need to let Auth0 know where we are allowed to send users after authentication or logout, These are the **Allowed Callback URLs** and the **Allowed Logout URLs**, in addition, if you would like Auth0 to handle refreshing the users token and session persistence, we will need to configure the **Allowed Web Origins** aswell.

Reviewer: Style: Exclamation Points

https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.1mrcu09

Reviewer: ```suggestion
One last step before we can begin integrating! We need to let Auth0 know where we are allowed to send users after authentication or logout, These are the **Allowed Callback URLs** and the **Allowed Logout URLs**, in addition, if you would like Auth0 to handle refreshing the users token and session persistence, we will need to configure the **Allowed Web Origins** as well.
```
typo

Reviewer: * Allowed Callback URL
* Allowed Logout URL
* Allowed Web Origins

Do you expect the reader to understand what each of the above are?  What values should I put in there?

If you expect the reader to understand, hyperlinks to reference documentation for anyone that is confused would help.  If you don't expect prior knowledge then you probably need to introduce each term

> We will set all three values here to `http://localhost:3000` for now (or whichever port you have configured your local React application runs on), In deployed applications, these values will be the fully qualified address of you applications (i.e. `https://example.com`)

Reviewer: This belongs with the last paragraph (may negate my previous comment)

> **Remember to Save your changes!**

Reviewer: Style: Exclamation Points

https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.1mrcu09

![Application URIs](/img/blog/integrating-auth0-with-react/application_uris.png){:width="426px" height="534px" loading="lazy"}

<br />

## Integrate with React

> > For this blog I will be working with **create-react-app** using the typescript template. The Auth0 SDK has full Typescript support in-built.

Reviewer: nitpick: Quote formatting - is that appropriate here?

Firstly, we will install the Auth0 SDK

```sh
npm install @auth0/auth0-react
```

The SDK has everything you will need to ingtegrate your application with Auth0, Under the hood it uses React Context to keep track of your users state.

The easiest way to integrate Auth0 is by wrapping your entire application with the convenient `Auth0Provider`

> ```tsx

Reviewer: You're assuming React knowledge here, which is fine, but context is useful.

Suggest adding the filename this snippet belongs to

Reviewer 2: I like adding file names to code snippets. Since Chris's blog highlight's "React" the assumption is safe.

import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <Auth0Provider
      domain='<Your Auth0 domain>'
      clientId='<Your applications Client ID>'
      redirectUri='http://localhost:3000'
    >
      <App />
    </Auth0Provider>
  </React.StrictMode>
);
```

> The `Auth0Provider` takes in two mandatory values, `domain` and `clientId`, Both of these can be found in your Settings of your Auth0 application. In addition, an optional but important parameter is the `redirectUri`, This will tell Auth0 where to send your users after authenticating successfully.

Reviewer: ```suggestion
The `Auth0Provider` takes in two mandatory values, `domain` and `clientId`, Both of these can be found in the Settings for your Auth0 application. In addition, an optional but important parameter is the `redirectUri`, This will tell Auth0 where to send your users after authenticating successfully.
```

The Auth0Provider keeps track of your users authenticated state and exposes all the Auth0 integrations to your application via React Context.

Congratulations! You've now integrated your Auth0 Application with your React Application!

### But wait, How do users login?

I knew we forgot something. OK, Let's give our users a login button.

```tsx

> // /src/components/LoginButton.tsx

Reviewer: I think there's a better way to highlight this.

`/src/components/LoginButton.tsx`:
```
code here
```
?

import React, { FC } from 'react';
import { useAuth0 } from "@auth0/auth0-react";

const LoginButton: FC = () => {
    const { loginWithRedirect } = useAuth0();

    const style = {width: '200px', height: '50px', margin: '20px'} // Give the button a bit of substance

    return <button style={style} onClick={() => loginWithRedirect()}>Log In with Auth0</button>;
}

export default LoginButton;
```

```tsx

> // /src/App.tsx

Reviewer: as above - doesn't matter which, but pick a style and stick with it.

...
import LoginButton from './components/LoginButton';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        ...
        <a
          ...
        >
          Learn React
        </a>
        <LoginButton />
      </header>
    </div>
  );
}
...
```

![Login Button](/img/blog/integrating-auth0-with-react/login_button.png){:width="450px" height="490px" loading="lazy"}

> The button itself makes use of one of the amazing Hooks provided by the Auth0 SDK `useAuth0`

Reviewer: does this need a link to the documentation?

`useAuth0` is our go to for a lot of functionality in our integration, But what we are interested in is the `loginWithRedirect` method.

This function will take our configuration of our Auth0Provider, and once executed will redirect the user to the **Auth0 Universal Login Page** and after a user has successfully authenticated, redirect them back to our application! Very cool!

Let's see what happens if we try to login.

![Universal Login Page](/img/blog/integrating-auth0-with-react/universal_login.png){:width="345px" height="555px" loading="lazy"}

> > Auth0 by default enabled Google as a social provider for all new applications, this connection uses some basic development credentials, so if you are looking to use Google as a social provider for production you should configure your own provider credentials in the Connections section of the Auth0 Application

Reviewer: nitpick: formatting

is quote formatting appropriate?

> Beautiful! Users can now Sign In, Sign Up and perform Password recoveries all with this universal login page!

Reviewer: Style: Exclamation Points

https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.1mrcu09

    return (
        <div>
            {isAuthenticated && user && (
                <>
                    <img src={user.picture} alt={user.name} />
                    <h2>{user.name}</h2>
                    <p>{user.email}</p>
                </>
            )}
        </div>
    );
}

export default Profile;
```

Once again, the `useAuth0` hook showing its strengths! We can get the user object, check if they are authenticated, and that everything has been fully loaded before displaying the user profile.

Let's add that to our homepage

```tsx
// /src/App.tsx
...
import LoginButton from './components/LoginButton';
import LogoutButton from './components/LogoutButton';
import Profile from './components/Profile';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <LoginButton />
        <LogoutButton />
        <Profile />
      </header>
    </div>
  );
}

export default App;
```

![Profile](/img/blog/integrating-auth0-with-react/profile.png){:width="580px" height="387px" loading="lazy"}

There we go! A fully integrated user Authenticated system and an amazing launch point for all of your Authentication and Authorization needs.

> ## Conclusion

Reviewer 2: A question I have is where do these records get stored (I'm assuming Auth0), and what security precautions do they undertake to keep things secure? Might'n be necessary for this blog, I'm just curious :)

Author: Yeah, Auth0 managed database, However they do have the ability to use your own self managed database if you want!

You can also use the same database accross multiple clients

Security wise, They have pretty solid [compliance](https://auth0.com/security) and honestly, I feel like it's even easier to understand how secure they are vs the Shared Responsibility model for Cognito. it's not hidden in whitepapers and random docs across the website

```tsx
// /src/index.tsx

import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <Auth0Provider
      domain='<Your Auth0 domain>'
      clientId='<Your applications Client ID>'
      redirectUri='http://localhost:3000'
    >
      <App />
    </Auth0Provider>
  </React.StrictMode>
);
```

The `Auth0Provider` takes in two mandatory values, `domain` and `clientId`, Both of these can be found in the Settings for your Auth0 application. In addition, an optional but important parameter is the `redirectUri`, This will tell Auth0 where to send your users after authenticating successfully.

The Auth0Provider keeps track of your users authenticated state and exposes all the Auth0 integrations to your application via React Context.

Congratulations! You've now integrated your Auth0 Application with your React Application!

### But wait, How do users login?

> I knew we forgot something. OK, Let's give our users a login button.

Reviewer 3: OK, let's give our users a login button.

Author: Updated

```

![Logout Button](/img/blog/integrating-auth0-with-react/logout.png){:width="275px" height="190px" loading="lazy"}

Awesome, we now have User Signup, Login, and Logout!

## User information and Profile

Now that our users can login, How do we go about getting their information such as Name, Email address, Profile Picture etc?

Alright, Let's create a new component that will retrieve that for us.

```tsx
// /src/components/Profile.tsx

import React, { FC } from 'react';
import { useAuth0 } from "@auth0/auth0-react";

const Profile: FC = () => {
    const { user, isAuthenticated, isLoading } = useAuth0();

    if (isLoading) {
        return <div>Loading...</div>
    }

    return (
        <div>
            {isAuthenticated && user && (
                <>
                    <img src={user.picture} alt={user.name} />
                    <h2>{user.name}</h2>
                    <p>{user.email}</p>
                </>
            )}
        </div>
    );
}

export default Profile;
```

Once again, the `useAuth0` hook showing its strengths! We can get the user object, check if they are authenticated, and that everything has been fully loaded before displaying the user profile.

> Let's add that to our homepage

Reviewer 3: perhaps another word other than Let's

Author: Good call, I overused that a bit at the end there

```tsx
// /src/App.tsx
...
import LoginButton from './components/LoginButton';
import LogoutButton from './components/LogoutButton';
import Profile from './components/Profile';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <LoginButton />
        <LogoutButton />
        <Profile />
      </header>
    </div>
  );
}

export default App;
```

![Profile](/img/blog/integrating-auth0-with-react/profile.png){:width="580px" height="387px" loading="lazy"}

There we go! A fully integrated user Authenticated system and an amazing launch point for all of your Authentication and Authorization needs.

## Conclusion

> We have just touched the surface of what Auth0 can do by implementing Authentication into our Application. Authorization is another piece of the puzzle that Auth0 can perform.

Author: That's the dream 😄 
Multi-tennancy and Authorization in the future
